drakrun/drakrun/storage.py

```diff
@@ -0,0 +1,264 @@
+
+        yield volume_path
+
+        subprocess.check_output(f"zfs destroy {tmp_snap}", shell=True)
+
+
+class Qcow2StorageBackend(StorageBackendBase):
+    """ Implements storage backend based on QEMU QCOW2 image format """
+
+    def __init__(self, install_info: InstallInfo):
+        super().__init__(install_info)
+        self.check_tools()
+
+    @staticmethod
+    def check_tools():
+        """ Verify existence of qemu-img """
+        try:
+            subprocess.check_output("qemu-img --version", shell=True)
+        except subprocess.CalledProcessError:
+            raise RuntimeError("Failed to determine qemu-img version. "
+                               "Make sure you have qemu-utils installed.")
+
+    def initialize_vm0_volume(self, disk_size: str):
+        try:
+            subprocess.check_output(
+                " ".join(
+                    [
+                        "qemu-img",
+                        "create",
+                        "-f",
+                        "qcow2",
+                        os.path.join(LIB_DIR, "volumes", "vm-0.img"),
+                        shlex.quote(disk_size),
+                    ]
+                ),
+                shell=True,
+            )
+        except subprocess.CalledProcessError:
+            raise RuntimeError("Failed to create a new volume using qemu-img.")
+
+    def snapshot_vm0_volume(self):
+        # We'll be using vm-0.img as backing storage
+        pass
+
+    def get_vm_disk_path(self, vm_id: int) -> str:
+        return f"tap:qcow2:{LIB_DIR}/volumes/vm-{vm_id}.img,xvda,w"
+
+    def rollback_vm_storage(self, vm_id: int):
+        volume_path = os.path.join(LIB_DIR, "volumes", f"vm-{vm_id}.img")
+        vm0_path = os.path.join(LIB_DIR, "volumes", "vm-0.img")
+        try:
+            os.unlink(volume_path)
+        except FileNotFoundError:
+            pass
+
+        subprocess.run(
+            [
+                "qemu-img",
+                "create",
+                "-f",
+                "qcow2",
+                "-o",
+                f"backing_file={vm0_path}",
+                volume_path
+            ],
+            check=True,
+        )
+
+    @contextlib.contextmanager
+    def vm0_root_as_block(self) -> Generator[str, None, None]:
+        try:
+            subprocess.check_output("modprobe nbd", shell=True)
+        except subprocess.CalledProcessError:
+            raise RuntimeError("Failed to load nbd kernel module")
+
+        try:
+            # TODO: this assumes /dev/nbd0 is free
+            subprocess.check_output(
+                f"qemu-nbd -c /dev/nbd0 --read-only {os.path.join(LIB_DIR, 'volumes', 'vm-0.img')}",
+                shell=True,
+            )
+        except subprocess.CalledProcessError:
+            raise RuntimeError("Failed to connect QCOW2 file to /dev/nbd0")
+
+        # we mount 2nd partition, as 1st partition is windows boot related and 2nd partition is C:\\
+        yield "/dev/nbd0p2"
```

Comment:
There is no guarantee at all that this will be 2nd partition, but that's just popular.

Reply:
Yes, there's also no guarantee that `ndb0` will be free to use, both of those are just assumptions.